File: trunk/src/tools/emutools/spcdasm.hs

import Text.ParserCombinators.Parsec hiding (spaces)
import Text.Printf

import Numeric

import Data.Maybe
import Data.List (unfoldr)
import Data.Bits
import Data.Word
import Data.Int
import Data.Char
import qualified Data.ByteString as BS
import qualified Data.IntMap as IM

import Control.Monad (liftM)

import Foreign (unsafePerformIO)
import System

-- Parser for TASM files

data Op = Op { opMnemonic :: String,
               opArgs :: String,
               opOpcode :: Int,
               opBytes :: Int,
               opRule :: String,
               opClass :: Int,
               opArgShift :: Int,
               opArgOr    :: Int                             
             } deriving (Eq, Ord, Show)


spaces = many (oneOf " \t")

pTable = do between (char '"') (char '"') (many (noneOf "\"")); newline
            xs <- many pTableEntry
            many comment
            return xs            
            
pTableEntry = do mnem <- many1 alphaNum;                  spaces
                 args <- many1 (satisfy (not . isSpace)); spaces
                 opcodeStr <- many1 hexDigit;             spaces
                 bytesStr <- many1 digit;                 spaces
                 rule <- many1 alphaNum;                  spaces
                 classStr <- many1 digit;                 spaces

                 argShift <- option 0 (liftM read (many1 digit)); spaces
                 argOr    <- option 0 (liftM (read . ("0x"++)) (many1 hexDigit))

                 newline
                 
                 return (Op { opMnemonic = mnem,
                              opArgs = args,
                              opOpcode = read ("0x"++opcodeStr),
                              opBytes = read bytesStr,
                              opRule = rule,
                              opClass = read classStr,
                              opArgShift = argShift,
                              opArgOr = argOr })
              <|> (comment >> pTableEntry)

comment = many (noneOf "\r\n") >> newline


-- Disassembler (doesn't support most TASM constructs)

optable = either (error . show) id (unsafePerformIO (parseFromFile pTable "tools/emutools/spc_asm.tab"))
opmap = IM.fromList [(opOpcode op, op) | op <- optable]

disasmOp :: BS.ByteString -> Int -> (String, Int)
disasmOp mem addr = (str, addr')
    where
    op = fromMaybe (Op { opMnemonic = "DB", opBytes = 1,
                         opArgs = "*", opRule = "T1", opClass = 1,
                         opArgShift = 0, opArgOr = 0xFF,
                         opOpcode = fromIntegral (mem `BS.index` addr) })
                   (IM.lookup (fromIntegral (mem `BS.index` addr)) opmap)

    addr' = addr + opBytes op
    str = printf "%04x: %6s %s" addr (opMnemonic op) (argStr (opArgs op) args)

    argStr "\"\""   []         = ""
    argStr ('*':ts) (arg:args) = printf "$%x" arg ++ argStr ts args
    argStr ('*':_)  []         = error ("not enough args for op " ++ show op)
    argStr (t:ts)   args       = t                :  argStr ts args
    argStr ""       []         = ""
    argStr ""       _          = error ("too many args ("++show args++") for op " ++ show op)

    args :: [Int]
    args | opBytes op == 1 =
             case opRule op of
             "NOP" -> []
             "T1" -> [(opOpcode op .&. opArgOr op) `shiftR` opArgShift op]
         | otherwise =
             map fromIntegral $
             case opRule op of
             "NOP" -> [baseArg]
             "CSWAP" -> [(baseArg `shiftR` 8) .&. 0xFF, baseArg .&. 0xFF]
             "CREL" -> [baseArg .&. 0xFF,
                        fromIntegral (addr + opBytes op + (fromIntegral (fromIntegral (baseArg `shiftR` 8) :: Int8)))]
             "R1" -> [fromIntegral (addr + opBytes op + (fromIntegral (fromIntegral baseArg :: Int8)))]
             rule -> error ("unimplemented rule "++rule)

    baseArg = realBaseArg `shiftR` opArgShift op

    realBaseArg :: Word
    realBaseArg = case opBytes op of
                  2 -> fromIntegral (BS.index mem (addr+1))
                  3 -> fromIntegral (BS.index mem (addr+1))
                       + (fromIntegral (BS.index mem (addr+2)) `shiftL` 8)
              

disasm :: BS.ByteString -> Int -> Int -> [String]
disasm mem start stop = unfoldr (\addr -> if addr < stop 
                                             then Just (disasmOp mem addr)
                                             else Nothing)
                                start

main = do [file] <- getArgs
          spcdata <- BS.readFile file
          let ram = BS.drop 0x100 spcdata
               
          putStr (unlines (disasm ram 0x0000 0x10000))

File: src/tools/emutools/spcdasm.py
import array

try:
    import psyco
except:
    pass

def uint8(n):
    n &= 255
    if n >= 128:
        n -= 256
    return n


rules = { "NOP":   lambda op, addr, n: [n],
          "CSWAP": lambda op, addr, n: [(n >> 8) & 255, n & 255],
          "CREL":  lambda op, addr, n: [(n & 255), addr + op.bytes + uint8(n >> 8)],
          "R1":    lambda op, addr, n: [addr + op.bytes + uint8(n >> 8)]
        }

class Op:
    def __init__(self, mnemonic, args, opcode, bytes, rule, klass, argShift="0", argOr="00"):
        self.mnemonic = mnemonic
        if args == '""':
            self.args = ''
        else:
            self.args = args
        self.opcode   = int(opcode, 16)
        self.bytes    = int(bytes)
        self.rule     = rules[rule]
        self.klass    = klass
        self.argShift = int(argShift)
        self.argOr    = int(argOr, 16)

ops = [None]*256

for line in file('tools/emutools/spc_asm.tab'):
    if line[0].isalnum():
        op = Op(*line.split())
        ops[op.opcode] = op
        

def disasm_op(mem, pc):
    def readint(addr, length):
        n = 0
        for i in range(length)[::-1]:
            n = n*256 + mem[addr+i]
        return n

    op = ops[mem[pc]]
    args = op.rule(op, pc, readint(pc+1, op.bytes-1))

    s = '%s ' % op.mnemonic

    for c in op.args:
        if c == '*':
            s += "$%x" % args.pop(0)
        else:
            s += c
    
    return (s, pc+op.bytes)

def disasm(mem, start, stop):
    pc = start
    while pc < stop:
        (op, newpc) = disasm_op(mem, pc)
        print "%04x: %-6s" % (pc, op)
        pc = newpc


if __name__ == '__main__':
    import sys
    mem = array.array('B', file(sys.argv[1]).read()[0x100:0x10100])
    disasm(mem, 0x0000, 0x10000)
